std/traits.d

```diff
@@ -8087,3 +8089,124 @@ enum isCopyable(S) = is(typeof(
     static assert(isCopyable!int);
     static assert(isCopyable!(int[]));
 }
+
+private template TypeOf(T...)
+    if (T.length == 1 && (is(T[0]) || is(typeof(T[0]))))
+{
+    static if (is(T[0]))
+        alias TypeOf = T[0];
+    else static if (is(typeof(T[0])))
+        alias TypeOf = typeof(T[0]);
+}
+
+@safe unittest
+{
+    assert(is(TypeOf!3 == int));
+    assert(is(TypeOf!int == int));
+    assert(is(TypeOf!([]) == void[]));
+    assert(!is(TypeOf!(mixin(__MODULE__))));
+    assert(is(TypeOf!TypeOf));
+}
+
+/**
+ * Get a list of the overloads for the given callable.
+ *
+ * Params:
+ *     func = The callable whose overload set to return.
+ * Returns:
+ *     A tuple of the overloads for the given callable.
+ */
+template getOverloads(func...)
+    if (func.length == 1 && isCallable!func)
+{
+    enum isModuleOrPackage(alias s) = !__traits(compiles, { alias _ = typeof(s); });
+    alias Parent = AliasSeq!(__traits(parent, func))[0];
+
+    static if (is(Parent) || isModuleOrPackage!Parent)
+        alias getOverloads = AliasSeq!(__traits(getOverloads, Parent, __traits(identifier, func)));
+    else
+        alias getOverloads = func;
+}
+
+version (unittest)
+{
+    // Local functions can't have overloads, so the functions used by
+    // getOverloads need to be defined here.
+    private:
+    int getOverloadsTest(int n = 3) { return n; }
+    int getOverloadsTest(int, int)  { return 3; }
+    int getOverloadsTest2(int, int) { return 3; }
+}
+
+///
+@safe unittest
+{
+    // Local functions can't have overloads, so these functions are defined above.
+    // int getOverloadsTest(int n = 3) { return n; }
+    // int getOverloadsTest(int, int)  { return 3; }
+    // int getOverloadsTest2(int, int) { return 3; }
+
+    assert(getOverloads!getOverloadsTest.length == 2);
+    assert(is(typeof(&getOverloadsTest) == typeof(&getOverloads!getOverloadsTest[0])));
+    assert(!is(typeof(&getOverloadsTest) == typeof(&getOverloads!getOverloadsTest[1])));
+    alias fn = getOverloadsTest;
+    assert(is(typeof(&fn) == typeof(&getOverloads!getOverloadsTest[0])));
+    assert(!is(typeof(&fn) == typeof(&getOverloads!getOverloadsTest[1])));
+    assert(getOverloads!getOverloadsTest2.length == 1);
+    assert(getOverloads!(() => 3).length == 1);
+    assert(getOverloads!(() => 3).length == 1);
+}
+
+/**
+ * Gets the overload for `func` that can be called with `Args` as arguments, if any.
+ *
+ * Params:
+ *    func = the overload set to search.
+ *    Args = the set of argument types to look for.
+ * Returns:
+ *    The overload that matches the argument types in Args
+ */
+template overloadFor(alias func, Args...)
+    if (isCallable!func)
+{
+    alias ArgTypes = staticMap!(TypeOf, Args);
```

Comment:
The docs say that `Args` is a set of types - this should be *sequence* of types. The docs don't mention also accepting values - is that something that `std.traits` has a precedent for? If so the docs for `Args` must mention values too.

If we want to support both types and values, maybe we only need to support one or the other, and not allow interleaving. In that case, we could remove the module symbol `TypeOf` and instead have:
```d
alias ArgTypes = is(Args) ? Args : typeof(Args);
```
If we do want interleaving, the `staticMap` expression could be rolled into an overload of `TypeOf` taking more than one argument.